File: Solutions/2. Add Two Numbers/Ruby/Add_Two_Numbers.rb
class ListNode
    attr_accessor :val, :next
    def initialize(val = 0, _next = nil)
        @val = val
        @next = _next
    end
end

def add_two_numbers(l1, l2)
    return add_two_numbers_with_carry(l1, l2, false)
end

def add_two_numbers_with_carry(l1, l2, c)
    if (l1 || l2 || c)
        sum = (l1&.val || 0) + (l2&.val || 0) + (c ? 1 : 0)
        return ListNode.new(sum % 10, add_two_numbers_with_carry(l1&.next, l2&.next, sum >= 10))
    end
end

n2 = ListNode.new(9);
n1 = ListNode.new(9, n2)
n0 = ListNode.new(9, n1)

m1 = ListNode.new(9)
m0 = ListNode.new(9, m1)

iter = add_two_numbers(n0, m0)

while iter
    puts iter.val
    iter = iter.next
end

File: Solutions/2. Add Two Numbers/JavaScript/Add_Two_Numbers.js
function ListNode(val, next) {
    this.val = (val===undefined ? 0 : val)
    this.next = (next===undefined ? null : next)
}

var addTwoNumbers = function(l1, l2) {
    return addTwoNumbersWithCarry(l1, l2, 0);
};

var addTwoNumbersWithCarry = function(l1, l2, carry) {
    if (!l1 && !l2 && !carry) {
        return null;
    }

    let num1 = 0;
    if (l1) {
        num1 = l1.val;
        l1 = l1.next;
    }

    let num2 = 0;
    if (l2) {
        num2 = l2.val;
        l2 = l2.next;
    }

    let s = (num1 + num2 + carry) % 10;
    let c = Math.floor((num1 + num2 + carry) / 10);

    return new ListNode(s, addTwoNumbersWithCarry(l1, l2, c));
};

var l1 = new ListNode(2, new ListNode(4, new ListNode(3)));
var l2 = new ListNode(5, new ListNode(6, new ListNode(4)));

console.log(addTwoNumbers(l1, l2));